packages/editor/src/components/provider/index.native.js

```diff
@@ -161,6 +163,11 @@ class NativeEditorProvider extends Component {
 			// purposes and will be replaced with actual logic in a later PR.
 			this.setState( { isHelpVisible: true } );
 		} );
+
+		// Request current block impressions from native app
+		requestBlockTypeImpressions( ( impressions ) => {
+			updateSettings( { impressions } );
+		}, NEW_BLOCK_TYPES );
```

Comment:
It feels like we should be specifying the starting value for the "impression count" for new blocks somewhere on the js-side, like here (or maybe it should be a part of `NEW_BLOCK_TYPES`), instead of doing it separately on [Android](https://github.com/wordpress-mobile/WordPress-Android/blob/0ef7ba7cfccd38dd99b10040fa78b61f856fffa0/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L3450) and [iOS](https://github.com/wordpress-mobile/WordPress-iOS/blob/d28d85ae0d96cd2423d7709addd357046c8b20fe/WordPress/Classes/ViewRelated/Gutenberg/GutenbergViewController.swift#L946). Is there any reason we would ever want Android and iOS to have different values?

Reply:
Or could we take that further and actually merge the new block types into the pre-existing impressions inside the callback? That way we would do the merging logic one time instead of separately on Android and iOS. The Android and iOS implementations would just become a simple key-value store.

In essence, I'm wondering if we would benefit from making it so the main apps just had an interface along these lines:

```
void getPersistedString(String key, persistenceCallback callback);
void persistString(String key, String value);
```

Then we would have a generic way to persist anything we want from Gutenberg (as long as we can convert it to/from a string) without ever having to touch code in the main apps. The only time we would need to go in and do the kind of native work you've had to do here would be when we need to also use the persisted value on the native side (because we're throwing away all type safety with this approach, and imo we only want to do that for things that are only being used within Gutenberg). 

I'm sure we could even avoid manually creating this interface by just using a RN library for Gutenberg-only persistence like this.

Reply:
On a semi-related note, this has got me wondering again whether we would giving the main apps a more generic interface for persisting and retrieving key-value pairs along these lines:

```
void getPersistedString(String key, Consumer<String> persistedStringCallback);
void persistString(String key, String value, PersistanceScope persistanceScope);
enum PersistenceScope { User, Global, Site }
```

That's not something we need to tackle in this PR though.

Reply:
> some form of key-value store does seem to be a very common requirement for many features. Building an abstraction to support this frequent requirement does make sense IMO. Surely at this point we have enough use cases to identify a common API for the temporary persistence layer JavaScript requires. Given the elongated timeline of this project, I would prefer to attempt that abstraction in a different PR/project.

👍  This does not necessarily need to be done as a part of this project.

Reply:
> > If I follow your proposal correctly, I originally opted against that approach to avoid a second trip over the native bridge, once to retrieve the persisted values and once to set the persisted values based upon NEW_BLOCK_TYPES configuration.

Could we make the extra trip a very rare occurrence? For example, I'm wondering if we could do something like:

1. Fetch impressions from native
2. On the JS side, check if the impressions include the blocks from `NEW_BLOCK_TYPES`. Only if the `NEW_BLOCK_TYPES` are missing do we "update" the impressions with the `NEW_BLOCK_TYPES` impressions (so this "second" call would only happen one time for each change to the `NEW_BLOCK_TYPES`). 

> Currently, the impressions are seeded and only set individually by name when a given block is inserted. This could be refactored so that the persisted value is set by passing a new object instead, but we would also need to immediately set the impressions within the request callback.

I think this could make sense, particularly with the flow I suggested above since it would avoid having to "set" the impressions for _each_ new block type. In fact, it might even enable us to avoid having any extra trips across the bridge--we could let the `NEW_BLOCK_TYPES` get "added" to the impression count on the native side as a part of the update that would occur whenever the number of impressions for one of the blocks gets updated (since we are now sending the entire "object," including the added-in-the-callback `NEW_BLOCK_TYPES` across with that update). 

The one concern with sending the full object that comes to mind for me is whether that object might grow over time to be so large that its size could impact performance (relative to just sending an update with a single key/value). That feels like a very hypothetical concern. I would love for us to have so many blocks that this object became too big, but I don't see it happening anytime in the remotely foreseeable future. Also, if we really wanted to, I guess we could avoid that size issue by just passing a map of updates instead of overwriting the entire object.
 
> Do you think deduplicating code is worth the second bridge trip?

I do, but I also think it is a point that reasonable people could disagree on. 🙂

Reply:
> Could we make the extra trip a very rare occurrence? 

@mchowning yes, I agree that a check within the callback for newly added block types could allow us to conditionally perform the second trip as needed. 

> In fact, [persisting the value with a new object] might even enable us to avoid having any extra trips across the bridge--we could let the `NEW_BLOCK_TYPES` get "added" to the impression count on the native side as a part of the update that would occur whenever the number of impressions for one of the blocks gets updated (since we are now sending the entire "object," including the added-in-the-callback `NEW_BLOCK_TYPES` across with that update).

Clever. I like it. I do have concern that it's possibly _too clever_, where it might cause confusion for a future contributor modifying this code, but I like it enough to explore. Hopefully I can identify a pattern/naming that makes it very clear what is occurring. 

> The one concern with sending the full object that comes to mind for me is whether that object might grow over time to be so large that its size could impact performance (relative to just sending an update with a single key/value). 

That makes sense. I agree this becoming a significant concern is likely a ways off. 

> > Do you think deduplicating code is worth the second bridge trip?
> 
> I do, but I also think it is a point that reasonable people could disagree on, so don't hesitate to let me know if you do. 🙂

I do not have a staunch opinion on the matter. I will explore postponing native persistence until a block is inserted. If that proves confusing, I will explore a conditional second bridge trip.

Reply:
>  I do have concern that it's possibly too clever, where it might cause confusion for a future contributor modifying this code

👍  Definitely a fair concern. It might be worth just having the occasional extra trip across the bridge when `NEW_BLOCK_TYPES` changes if that allows us to make the code easier to understand.

Reply:
I was thinking about this last night, and one additional reason to favor moving logic to js from the native side (apart from avoiding duplicated logic) is that it would make it easier for other apps to integrate gb-mobile. There will be less complexity they will need to add if we keep as much complexity as possible "inside" gb-mobile.